lambdas/s3hash/src/t4_lambda_s3hash/__init__.py

```diff
@@ -222,47 +218,110 @@ async def upload_part(mpu: MPURef, src: S3ObjectSource, etag: str, part: PartDef
         # otherwise we can end up with invalid checksum.
         CopySourceIfMatch=etag,
     )
-    cs = res["CopyPartResult"].get("ChecksumSHA256")  # base64-encoded
-    assert cs is not None
-    return base64.b64decode(cs)
+    copy_result = res["CopyPartResult"]
+    assert "ETag" in copy_result
+    assert "ChecksumSHA256" in copy_result
+    return PartUploadResult(
+        etag=copy_result["ETag"],
+        sha256=copy_result["ChecksumSHA256"],
+    )
 
 
-async def compute_part_checksums(
+async def upload_parts(
     mpu: MPURef,
     location: S3ObjectSource,
     etag: str,
     parts: T.Sequence[PartDef],
-) -> T.List[bytes]:
+) -> T.List[PartUploadResult]:
     s = asyncio.Semaphore(MPU_CONCURRENCY)
 
     async def _upload_part(*args, **kwargs):
         async with s:
             return await upload_part(*args, **kwargs)
 
     uploads = [_upload_part(mpu, location, etag, p) for p in parts]
-    checksums: T.List[bytes] = await asyncio.gather(*uploads)
+    return await asyncio.gather(*uploads)
+
+
+async def compute_part_checksums(
+    mpu: MPURef,
+    location: S3ObjectSource,
+    etag: str,
+    parts: T.Sequence[PartDef],
+) -> T.List[bytes]:
+    part_upload_results = await upload_parts(
+        mpu,
+        location,
+        etag,
+        parts,
+    )
+    checksums = [base64.b64decode(part_upload_result.sha256) for part_upload_result in part_upload_results]
     return checksums
 
 
 MPU_DST = S3ObjectDestination(bucket=SERVICE_BUCKET, key=SCRATCH_KEY_SERVICE)
 
 
+class PartUploadResult(pydantic.BaseModel):
+    etag: str
+    sha256: str  # base64-encoded
+
+    @property
+    def boto_args(self):
+        return {
+            "ETag": self.etag,
+            "ChecksumSHA256": self.sha256,
+        }
+
+
+class MPURef(MPURefBase):
+    _completed: bool = pydantic.PrivateAttr(default=False)
+
+    @property
+    def completed(self):
+        return self._completed
+
+    async def complete(self, parts: T.Sequence[PartUploadResult]):
+        if self.completed:
```

Comment:
seems like this should be a promise-like thing, or at least a three-state thing (true/false/pending), otherwise we risk running into race conditions

Reply:
or maybe even a four-state thing, i.e. init / pending / completed / aborted 🤔

Reply:
> seems like this should be a promise-like thing, or at least a three-state thing (true/false/pending), otherwise we risk running into race conditions

can that really happen with asyncio?

Reply:
well, seems like in your design there's no places where a race can occur